File: templates/problems/05_binary_search/03_koko_bananas/python/solution.py
import math


def solve_min_eating_speed(piles: list[int], h: int) -> int:
    """Binary search for the minimum eating speed.

    Time: O(n log m), Space: O(1), where m is the max pile size.
    """
    lo, hi = 1, max(piles)
    while lo < hi:
        mid = lo + (hi - lo) // 2
        hours = sum(math.ceil(p / mid) for p in piles)
        if hours <= h:
            hi = mid
        else:
            lo = mid + 1
    return lo

File: templates/problems/05_binary_search/03_koko_bananas/typescript/solution.ts
export function solveMinEatingSpeed(piles: number[], h: number): number {
  // Binary search for the minimum eating speed.
  // Time: O(n log m), Space: O(1), where m is the max pile size.
  let lo = 1;
  let hi = Math.max(...piles);
  while (lo < hi) {
    const mid = lo + Math.floor((hi - lo) / 2);
    const hours = piles.reduce((sum, p) => sum + Math.ceil(p / mid), 0);
    if (hours <= h) {
      hi = mid;
    } else {
      lo = mid + 1;
    }
  }
  return lo;
}
